bytestring/common/src/ByteString.kt

```diff
@@ -0,0 +1,452 @@
+     * [startIndex] and ending at [endIndex].
+     *
+     * @param startIndex the start index (inclusive) of a subsequence to copy.
+     * @param endIndex the end index (exclusive) of a subsequence to copy, [size] be default.
+     *
+     * @throws IndexOutOfBoundsException when [startIndex] or [endIndex] is out of range of byte string indices.
+     * @throws IllegalArgumentException when `startIndex > endIndex`.
+     */
+    public fun substring(startIndex: Int, endIndex: Int = size): ByteString = if (startIndex == endIndex) {
+        EMPTY
+    } else {
+        ByteString(data, startIndex, endIndex)
+    }
+
+    /**
+     * Compares a byte sequence wrapped by this byte string to a byte sequence wrapped by [other]
+     * in lexicographical order.
+     * Byte values are compared as unsigned integers.
+     *
+     * @param other the byte string to compare this string to.
+     */
+    override fun compareTo(other: ByteString): Int {
+        if (other === this) return 0
+
+        for (i in 0 until min(size, other.size)) {
+            val cmp = data[i].toUByte().compareTo(other[i].toUByte())
+            if (cmp != 0) return cmp
+        }
+
+        return size.compareTo(other.size)
+    }
+
+    /**
+     * Returns a string representation of this byte string. A string representation consists of [size] and
+     * a hexadecimal-encoded string of a byte sequence wrapped by this byte string.
+     *
+     * The string representation has the following format `ByteString(size=3 hex=ABCDEF)`,
+     * for empty strings it's always `ByteString(size=0)`.
+     *
+     * Note that a string representation includes the whole byte string content encoded.
+     * Due to limitations exposed for the maximum string length, an attempt to return a string representation
+     * of too long byte string may fail.
+     */
+    override fun toString(): String {
+        if (isEmpty()) {
+            return "ByteString(size=0)"
+        }
+        // format: "ByteString(size=XXX hex=YYYY)"
+        val sizeStr = size.toString()
+        val len = 22 + sizeStr.length + size * 2
+        return with(StringBuilder(len)) {
+            append("ByteString(size=")
+            append(sizeStr)
+            append(" hex=")
+            for (i in 0 until size) {
+                val b = this@ByteString[i].toInt()
+                append(HEX_DIGITS[(b ushr 4) and 0xf])
+                append(HEX_DIGITS[b and 0xf])
+            }
+            append(')')
+        }.toString()
+    }
+
+    /**
+     * Returns a reference to the underlying array.
+     *
+     * These methods return reference to the underlying array, not to its copy.
+     * Consider using [toByteArray] if it's impossible to guarantee that the array won't be modified.
+     */
+    internal fun getBackingArrayReference(): ByteArray = data
+}
+
+/**
+ * Returns the range of valid byte indices for this byte string.
+ */
+public fun ByteString.indices(): IntRange = 0 until size
```

Comment:
`indicies` rather should be a property than a method, as it is in other array-like structures